### PR TITLE
Refs [TASK][26876]: Finishing use-cases SearchFragment and DetailFragment

### PR DESCRIPTION
**Pull request by aug16vcc**
## Related Tickets
- [#Ticket ID](https://edu-redmine.sun-asterisk.vn/issues/26876)

## WHAT
-  Finishing use-cases SearchFragment and DetailFragment

## HOW
-  Finishing use-cases SearchFragment and DetailFragment
- Refactoring code
- Format code

## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Vu Chi Cong | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
Conventions | Do the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [ ] no</li> 
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/boringweather/base/BaseFragment.kt [hunk 1]
```diff
@@ -7,7 +7,7 @@ import android.view.ViewGroup
 import androidx.annotation.LayoutRes
 import androidx.fragment.app.Fragment
 
-abstract class BaseFragment: Fragment() {
+abstract class BaseFragment : Fragment() {
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/boringweather/base/BaseFragment.kt:10]:
```suggestion
abstract class BaseFragment: Fragment() {
```

File: app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt [hunk 2]
```diff
@@ -21,55 +40,120 @@ class SearchFragment : BaseFragment(), SearchContract.View {
...
+    private lateinit var fusedLocationClient: FusedLocationProviderClient
+
+    private val cityAdapter: CityAdapter by lazy {
+        CityAdapter(
+            onclickListener = {
+                if (defaultSharedPreferences.selectedCityId == City.default.id)
+                    toggleSelectedCity(it)
+                sendFragmentResult(it)
+                findNavigator()?.popBackStack()
+            }, onBookMarkListener = {
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt]:
```suggestion
            },
            onBookMarkListener = {
```

File: app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt [hunk 3]
```diff
@@ -21,55 +40,120 @@ class SearchFragment : BaseFragment(), SearchContract.View {
...
+
+    private val pendingRestoreRecyclerViewPosition = Single<Int>()
+
+    private lateinit var fusedLocationClient: FusedLocationProviderClient
+
+    private val cityAdapter: CityAdapter by lazy {
+        CityAdapter(
+            onclickListener = {
+                if (defaultSharedPreferences.selectedCityId == City.default.id)
+                    toggleSelectedCity(it)
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt]:
```suggestion
                if (defaultSharedPreferences.selectedCityId == City.default.id) {
                    toggleSelectedCity(it)
                }
```

File: app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt [hunk 4]
```diff
@@ -21,55 +40,120 @@ class SearchFragment : BaseFragment(), SearchContract.View {
...
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
 
         initView()
         initListener()
-        requestInputFocus(editTextSearch)
+        if (context?.networkState == null) {
+            context?.defaultSharedPreferences?.lastSearchInput = ""
+            showError(R.string.error_network_no_connection)
+        } else requestInputFocus(editTextSearch)
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt]:
```suggestion
        } else {
            requestInputFocus(editTextSearch)
        }
```

File: app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt [hunk 5]
```diff
@@ -21,55 +40,120 @@ class SearchFragment : BaseFragment(), SearchContract.View {
...
+    }
+
+    private fun toggleStartIcon() {
+        if (defaultSharedPreferences.selectedCityId == City.default.id) {
+            textInputLayoutSearch.setStartIconDrawable(R.drawable.ic_round_help_outline_24)
+        } else {
+            textInputLayoutSearch.setStartIconDrawable(R.drawable.ic_round_arrow_back_24)
+            // textInputLayoutSearch.startIconDrawable =
+            //     AppCompatResources.getDrawable(requireContext(), R.drawable.ic_round_arrow_back_24)
+            //         ?.also { DrawableCompat.setTint(it, R.attr.colorPrimaryDark) }
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt:156]: remove comment

File: app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt [hunk 6]
```diff
@@ -21,55 +40,120 @@ class SearchFragment : BaseFragment(), SearchContract.View {
...
+    private fun toggleEndIcon(input: String) = input.isEmpty().let { empty ->
+        textInputLayoutSearch.endIconDrawable = ContextCompat.getDrawable(
+            requireContext(),
+            if (empty) R.drawable.ic_round_my_location_24
+            else R.drawable.ic_round_cancel_24
+        )
+        textInputLayoutSearch.suffixText =
+            if (empty) getString(R.string.title_search_use_device_location)
+            else getString(R.string.title_search_clear)
+
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt:151]: remove blank line

File: app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt (hunk 5 above)
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt]: remove commented code

File: app/src/main/res/layout/fragment_detail.xml [hunk 7]
```diff
@@ -10,7 +10,16 @@
         android:id="@+id/motionDetail"
         android:layout_width="match_parent"
         android:layout_height="match_parent"
-        app:layoutDescription="@xml/fragment_detail_scene">
+        app:layoutDescription="@xml/fragment_detail_scene"
+        app:motionDebug="SHOW_PATH">
+
```
Comment (tuanpa-0908) [inline app/src/main/res/layout/fragment_detail.xml:14]: rb

File: app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt [hunk 8]
```diff
@@ -21,55 +40,120 @@ class SearchFragment : BaseFragment(), SearchContract.View {
...
+            ?.let { editTextSearch.text = it }
+
+        toggleStartIcon()
     }
 
     private fun initListener() {
-        textInputLayoutSearch.setStartIconOnClickListener { findNavigator()?.popBackStack() }
+        textInputLayoutSearch.setStartIconOnClickListener {
+            if (defaultSharedPreferences.selectedCityId != City.default.id) findNavigator()?.popBackStack()
+            else showError(R.string.error_search_need_select_city)
```
Comment (thonc-2135) [inline app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt]: nên có {}

File: app/src/main/res/layout/item_weather_hourly_expandable.xml [hunk 9]
```diff
@@ -45,10 +46,35 @@
...
+        app:layout_constraintTop_toBottomOf="@+id/textTemperature" />
+
+    <TextView
+        android:id="@+id/textTemperature"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginEnd="@dimen/dp_24"
+        android:drawablePadding="@dimen/dp_8"
+        android:text="@string/title_holder_temperature"
+        android:textSize="35sp"
```
Comment (tuanpa-0908) [inline app/src/main/res/layout/item_weather_hourly_expandable.xml]: hard code

File: app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt (hunk 6 above)
Comment (thonc-2135) [inline app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt:151]: remove blank

File: app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt [hunk 10]
```diff
@@ -78,19 +162,105 @@ class SearchFragment : BaseFragment(), SearchContract.View {
...
-        context?.defaultSharedPreferences?.selectedCityId = cityId
+    private fun searchUsingDeviceLocation() {
+        val permissions = arrayOf(
+            Manifest.permission.ACCESS_FINE_LOCATION,
+            Manifest.permission.ACCESS_COARSE_LOCATION
+        )
+        val requestPermissionLauncher =
+            registerForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) { isGranted ->
+                if (isGranted.all { it.value }) performSearchUsingDeviceLocation()
+                else showError(R.string.error_search_location_permission_denied)
```
Comment (thonc-2135) [inline app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt]: thêm {} nếu như k thể viết hết if else vào 1 dòng thì phải

File: app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt [hunk 11]
```diff
@@ -78,19 +162,105 @@ class SearchFragment : BaseFragment(), SearchContract.View {
...
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun performSearchUsingDeviceLocation() {
+        context?.locationManager?.let {
+            if (LocationManagerCompat.isLocationEnabled(it)) {
+                fusedLocationClient.lastLocation.addOnSuccessListener { location ->
+                    if (location != null) {
+                        presenter?.searchCityByLocation(location)
+                    } else showError(R.string.error_search_location_null)
```
Comment (thonc-2135) [inline app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt]: ngoặc kép {}

File: app/src/main/res/layout/item_weather_hourly_expandable.xml [hunk 12]
```diff
@@ -1,15 +1,17 @@
 <?xml version="1.0" encoding="utf-8"?>
 <androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
     android:layout_width="match_parent"
     android:layout_height="wrap_content"
-    android:background="@drawable/background_ripple">
+    android:background="@drawable/background_ripple"
+    tools:layout_height="400dp">
```
Comment (thonc-2135) [inline app/src/main/res/layout/item_weather_hourly_expandable.xml]: hard code, bỏ luôn tools đi

File: app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt (hunk 8 above)
Comment (huytq-2422) [reply app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt]: Conventions đây nhé 
https://developer.android.com/kotlin/style-guide#braces

File: app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt (hunk 10 above)
Comment (huytq-2422) [reply app/src/main/java/com/sunasterisk/boringweather/ui/search/SearchFragment.kt]: https://developer.android.com/kotlin/style-guide#braces

File: app/src/main/res/layout/fragment_detail.xml (hunk 7 above)
Comment (aug16vcc) [reply app/src/main/res/layout/fragment_detail.xml:14]: rb  là gì vậy?

Comment (thonc-2135) [reply app/src/main/res/layout/fragment_detail.xml:14]: remove blank . 22 pull vẫn chưa đoán được à =))

**Pull request merged**